spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteMergeIntoTable.scala

```diff
@@ -0,0 +1,267 @@
+          }
+
+          m.copy(rewritePlan = Some(rewritePlan))
+
+        case p =>
+          throw new AnalysisException(s"$p is not an Iceberg table")
+      }
+  }
+
+  // build a rewrite plan for sources that support replacing groups of data (e.g. files, partitions)
+  private def buildReplaceDataPlan(
+      relation: DataSourceV2Relation,
+      table: RowLevelOperationTable,
+      source: LogicalPlan,
+      cond: Expression,
+      matchedActions: Seq[MergeAction],
+      notMatchedActions: Seq[MergeAction]): ReplaceData = {
+
+    // resolve all needed attrs (e.g. metadata attrs for grouping data on write)
+    val metadataAttrs = resolveRequiredMetadataAttrs(relation, table.operation)
+
+    // construct a scan relation and include all required metadata columns
+    val readRelation = buildReadRelation(relation, table, metadataAttrs)
+    val readAttrs = readRelation.output
+
+    // project an extra column to check if a target row exists after the join
+    // project a synthetic row ID to perform the cardinality check
+    val rowFromTarget = Alias(TrueLiteral, ROW_FROM_TARGET)()
+    val rowId = Alias(MonotonicallyIncreasingID(), ROW_ID)()
+    val targetTableProjExprs = readAttrs ++ Seq(rowFromTarget, rowId)
+    val targetTableProj = Project(targetTableProjExprs, readRelation)
+
+    // project an extra column to check if a source row exists after the join
+    val rowFromSource = Alias(TrueLiteral, ROW_FROM_SOURCE)()
+    val sourceTableProjExprs = source.output :+ rowFromSource
+    val sourceTableProj = Project(sourceTableProjExprs, source)
+
+    // use left outer join if there is no NOT MATCHED action, unmatched source rows can be discarded
+    // use full outer join in all other cases, unmatched source rows may be needed
+    // disable broadcasts for the target table to perform the cardinality check
+    val joinType = if (notMatchedActions.isEmpty) LeftOuter else FullOuter
+    val joinHint = JoinHint(leftHint = Some(HintInfo(Some(NO_BROADCAST_HASH))), rightHint = None)
+    val joinPlan = Join(targetTableProj, sourceTableProj, joinType, Some(cond), joinHint)
+
+    val rowIdAttr = ExtendedV2ExpressionUtils.resolveRef[AttributeReference](
+      FieldReference(ROW_ID),
+      joinPlan)
+    val rowFromSourceAttr = ExtendedV2ExpressionUtils.resolveRef[AttributeReference](
+      FieldReference(ROW_FROM_SOURCE),
+      joinPlan)
+    val rowFromTargetAttr = ExtendedV2ExpressionUtils.resolveRef[AttributeReference](
+      FieldReference(ROW_FROM_TARGET),
+      joinPlan)
+
+    // add an extra matched action to output the original row if none of the actual actions matched
+    // this is needed to keep target rows that should be copied over as we are working with groups
+    val mergeRowsParams = MergeRowsParams(
+      isSourceRowPresent = IsNotNull(rowFromSourceAttr),
+      isTargetRowPresent = IsNotNull(rowFromTargetAttr),
+      matchedConditions = matchedActions.map(actionCondition) :+ TrueLiteral,
```

Comment:
Why is `rewriteMatchedActions` no longer needed in this version? Is that because this may not be optimized yet and the condition could be `1 = 1` instead of a `true` literal?

Reply:
Oh, I see. That's why there's an extra `true` here and why it projects `readAttrs` below. Maybe add a comment to explain or fix up the outputs/conditions in a method like before.

Reply:
And I also saw the later comment that the optimizer doesn't run on the expressions. That's concerning.

Reply:
Yeah, it is a little bit tricky. I do have a comment above but let me see what can be done to simplify this place.

Reply:
I've restructured this place a bit and added that comment right next to it. Let me know if that's better.